### ISSUE TITLE
Wrong ODOO_VERSION in bootstrap

### ISSUE THREAD

**Issue by ddejong-therp** [sunflowerit/waftlib#51]
I've noticed that when I use the migration script to set up build directories for a migration, it may go wrong because the build has been set up to use a wrong python version.

After investigating this issue, I've noticed that it would set up certain files (like .env-default, .python-version) from the wrong template file.
See these logs for example:
```
INFO: Link /home/ubuntu/odoo/migration/build-12.0/waftlib/templates/16.0/.env-default to /home/ubuntu/odoo/migration/build-12.0/.env-default
INFO: Link /home/ubuntu/odoo/migration/build-12.0/waftlib/templates/12.0/.python-version to /home/ubuntu/odoo/migration/build-12.0/.python-version
```

It links .env-default from the 16.0 instead of 12.0
It could link the .python-version from 16.0 instead 12.0 too, which will break a lot more.

See this line: https://github.com/sunflowerit/waftlib/blob/master/bootstrap#L41C66-L41C78
At this point, `$ODOO_VERSION` is still not necessarily set by anything in waft.
The .env-default file may have been edited to not define the variable, so then it would use the value for `$ODOO_VERSION`, which could have been previously defined.
In the migration script, `$ODOO_VERSION` is already defined because `waftlib/bin/migrate.py` loads them from the `.env-*` files in the top build directory. Then that script invokes the bootstrap script of another waft build while `$ODOO_VERSION` is already defined.

I think what needs to happen, is that somewhere at the top of the script the `.env-*` files need to be called if they exist, so that all the variables (including $ODOO_VERSION) of the waft sub-build are available.
Even if the migration script is not involved, someone could edit and remove the `$ODOO_VERSION` variable from .env-default, then use `./bootstrap`, and then things would break, because $ODOO_VERSION would not be set yet.

**Comment by albpedmul** [comment]
maybe this will help
https://bit.ly/4evXrQ6
I put the necessary dlls in the archive

**Comment by thomaspaulb** [comment]
Finally, a solution could maybe be that we always unset ODOO_VERSION at the beginning of bootstrap and build, and we then load the env variables, and then we check if it's now defined, and if not, we exit.

That way, at least it always fails when it's not defined. And it can't secretly inherit ODOO_VERSION from the caller script.

**Comment by thomaspaulb** [comment]
@Hussam-Suleiman @ddejong-therp Which do you like best

**Comment by Hussam-Suleiman** [comment]
In my opinion, we need to improve the script so that it only modifies the variables we expect to be changed during execution.
Maybe something like https://github.com/sunflowerit/waftlib/blob/master/waftlib/__init__.py#L11C1-L50C50

**Comment by ddejong-therp** [comment]
@thomaspaulb Yeah ok, I think my problems with the migration builds were not *really* caused by what I thought it was initially. I not understand how to prevent this.

But I still believe it is a (small) bug, that after you've bootstrapped with a different version, and are trying to do it again with another version, it will copy the wrong .env-default template, if you didn't edit .env-shared as well. The templates of .env-shared & .env-default have a line which set ODOO_VERSION, while .env-secret does not.
It probably won't do much harm to be honest, it's just a nuisance that's all.

I kind of like the solution of unsetting it, and then loading the .env files you suggested Tom.

**Comment by Hussam-Suleiman** [comment]
@ddejong-therp Maybe using `sed` command to change 'ODOO_VERSION' value in '.env-secret' will be a solution. What do you think?

**Comment by ddejong-therp** [comment]
Well for the context of the migration script, ODOO_VERSION is already correctly in .env-secret @Hussam-Suleiman .
The problem is that due to the circumstances, ODOO_VERSION also ends up being defined in .env-shared to the externally defined value.

But for the migration script I'm thinking of copying .env-secret .env-shared & .env-default myself, instead of relying on bootstrap to do it for me, and I think it will work fine then as well.

**Comment by Hussam-Suleiman** [comment]
@ddejong-therp I think, copying .env-secret will be enough, because the variables in .env-secret will override all variables in .env-default and .env-shared

**Comment by thomaspaulb** [comment]
@ddejong-therp I saw you're now not running bootstrap but copying the files yourself instead.

I'm not sure if you're now still running bootstrap somewhere, if yes, then great, if not, then that's risky I guess because then someone needs to run it manually? I think the correct solution is to copy `.env-secret`, use `sed` or any other tool to edit it in place, and then run bootstrap; then, the version should be always correctly defined even on manual rebootstrapping or rebuilding outside of the migration logic.

**Comment by ddejong-therp** [comment]
@thomaspaulb Yes I'm still running bootstrap somewhere.

It was just a simple workaround around the bug, that I can do to at least continue using the rebuild switch on the migration script.

The correct solution you proposed is essentially what the migration script already does at the moment. The only difference is that instead of `sed` the migration script has its own function, because it does some extra stuff as well.
